src/composite.c

```diff
@@ -72,40 +82,22 @@ int Composite(VipsObject *context, VipsImage *src, VipsImage *dst, VipsImage **o
   //
   // Wikipedia:
   // out_rgb = (src_rgb * src_a + dst_rgb * dst_a * (1 - src_a)) / out_a
+  //                                                ^^^^^^^^^^^
+  //                                                    t0
   //
-  // `vips_ifthenelse` with `blend=TRUE`: http://bit.ly/1KoSsga
-  // out = (cond / 255) * in1 + (1 - cond / 255) * in2
-  //
-  // Substitutions:
+  // Omit division by `out_a` since `Compose` is supposed to output a
+  // premultiplied RGBA image as reversal of premultiplication is handled
+  // externally.
   //
-  //     cond --> src_a
-  //     in1 --> src_rgb
-  //     in2 --> dst_rgb * dst_a (premultiplied destination RGB)
-  //
-  // Finally, manually divide by `out_a` to unpremultiply the RGB channels.
-  // Failing to do so results in darker than expected output with low
-  // opacity images.
-  //
-  VipsImage *dstRGBPremultiplied;
-  if (vips_multiply(dstRGB, dstAlphaNormalized, &dstRGBPremultiplied, NULL))
-    return -1;
-
-  vips_object_local(context, dstRGBPremultiplied);
-
+  VipsImage *t2;
   VipsImage *outRGBPremultiplied;
-  if (vips_ifthenelse(srcAlpha, srcRGB, dstRGBPremultiplied,
-      &outRGBPremultiplied, "blend", TRUE, NULL))
+  if (vips_multiply(dstRGBPremultiplied, t0, &t2, NULL) ||
+      vips_add(srcRGBPremultiplied, t2, &outRGBPremultiplied, NULL))
     return -1;
 
+  vips_object_local(context, t2);
   vips_object_local(context, outRGBPremultiplied);
```

Comment:
@lovell: Isn’t this another memory leak: When `vips_multiply` succeeds `t2` is allocated, but then `vips_add` fails and the function returns without ever calling `vips_object_local(context, t2)`. I originally combined multiple operations into one—`if (vips_* || vips_* || vips_*)`—based on @jcupitt’s pattern using `vips_object_local_array`: https://gist.github.com/jcupitt/abacc012e2991f332e8b#file-composite2-c-L16-L22. However in his case, all `t[…]` seemed to be pre-hooked onto `context` right away. No more extra `vips_object_local` calls afterwards. Then we switched back to using individual variables for better readability but I kept the `if (… || … || …)` pattern. Your code uses individual variables but only single operations per `if (…)`. Am I right in identifying this as a potential leak (in error cases) or did I miss something?